security-controls/sast.py

```diff
@@ -0,0 +1,311 @@
+    '''path for codacy org'''
+    return f'{CODACY_BASE}/organizations/gh/{organization}'
+
+
+repo_re = re.compile(r'https://(www.)?github.com/([a-zA-Z0-9\-_]+)/([a-zA-Z0-9\-_]+)')
+RE_GROUP_ORG = 2
+RE_GROUP_REPO = 3
+
+SONAR_ORGS = {
+    "broadinstitute": "dsp-appsec",
+    "DataBiosphere": "broad-databiosphere",
+    "CancerDataAggregator": "cancerdataaggregator"
+}
+
+CODACY_ORGS = [
+    "DataBiosphere",
+    "broadinstitute"
+]
+
+CODACY = "Codacy"
+SONAR = "SonarCloud"
+GITHUB = "GitHub"
+
+RepoKey = tuple[str, str] # organization, repository
+Repo = dict[str, object] # key is source e.g. CODACY, GITHUB; value is repo info from source
+Repos = dict[RepoKey, Repo]
+CodacyOrgData = dict[str, list] # {"org": [ user_suggestion, ... ]}
+
+def get_if(obj: dict, key: str) -> str:
+    '''return obj[value] or empty string if missing'''
+    value = obj.get(key)
+    if value is None:
+        value = ""
+    return value
+
+def repo_list_from_security_controls(repos: Repos):
+    """Initalize repos from Firestore."""
+    sc_collection = fs.collection(SECURITY_CONTROLS)
+    sc_docs = sc_collection.stream()
+    for doc in sc_docs:
+        sc_record = doc.to_dict()
+        repo_url = get_if(sc_record,'github')
+        match = repo_re.match(repo_url)
+        if match is None:
+            logging.warning('SAST controls ignoring %s github="%s"', doc.id, repo_url)
+            continue
+        org = match.group(RE_GROUP_ORG)
+        repo_name = match.group(RE_GROUP_REPO)
+        if org.lower() == 'databiosphere':
+            org = 'DataBiosphere'
+        if org.lower() == 'broadinstitute':
+            org = 'broadinstitute'
+        repo = get_repo(repos, (org, repo_name))
+        sc_record[ID] = doc.id
+        repo[SC_PREVIOUS] = sc_record
+        logging.debug('SAST controls to be updated for %s github="%s"', doc.id, repo_url)
+
+
+def list_github(repos: Repos):
+    '''Get GitHub metadata for all repos.'''
+    count = 0
+    for org, repo_name in repos.keys():
+        list_repo_info(repos, org, repo_name)
+        count += 1
+        if count > 500:
+            break
+
+def quote(name: str) -> str:
+    '''Enclose string in escaped double quotes.'''
+    return r'\"' + name + r'\"'
+
+def body(gql: str, params: dict) -> str:
+    '''Expand parameters and wrap gql in JSON.'''
+    gql = gql.replace('\n',' ')
+    for key, value in params.items():
+        gql = gql.replace("$" + key, quote(value))
+    return '{ "query": "' + gql + '" }'
+
+def list_repo_info(repos: Repos, org, repo_name):
+    '''Pull GitHub metadata on this repo.'''
+
+    logging.debug("GitHub %s %s", repo_name, org)
+    headers = {
+        "Authorization": f"bearer {GITHUB_TOKEN}"
+    }
+
+    gql = '''
+            query {
+                repository(name: $repo_name, owner: $org) {
+                    visibility
+                    isArchived
+                    primaryLanguage {
+                        name
+                    }
```

Comment:
Why is this necessary for SAST?

Reply:
SAST policy and tool support depends on visibility and code language.